fix overlapping block Jacobi

jgedicke opened the pull request:
I found this bug while using overlapping block Jacobi preconditioners, originally in the relaxation class, in combination with MG.
The mathematical background is that the overlapping Jacobi preconditioner can be viewed as a additive Schwarz method with the emphasis to the word additive!!!
The logic of line 231 in relaxation_block.templates.h is that we add (!!!) the corrections to the previous iterate. Note that we initialize dst always with the previous iterate, thus we do only need to add (!!!) the corrections here.
So why did it work before for the other precontitioners, SOR, SSOR. If we look at the methods we see that for both SOR and SSOR we do iterate on a single vector dst (which is exactly what you have to do for a Gauss-Seidel loop), so then line  231 in relaxation_block.templates.h reads in that case with dst,prev pointing to the same memory `dst(j) = dst(j) + z` for some double z. Changing that to `dst(j) += z` is clearly the same in that case.
Now what about Jacobi then. Ok we do initialize dst with the old iterate so that we compute by the new line all together dst = old + sum_{blocks} A^{-1} residual, where residual is computed by the previous (old) vector. Before we where storing (not adding) a single value in dst(j) = old(j) + Correction_j, which is correct if and only if the blocks are not overlapping. In case of overlapp, one has to add them together as already explained.
In particular I also made the changes in precondition_block.templates.h so that we always start with the old vector in dst also in Jacobi and then change the two lines in forward_step and backward_step to (sorry to say it again) **add** the corrections to the old vector.
Best regards

bangerth commented:
I'll let others comment on the merits of this as I suspect that there are others who understand these things better than I do. But I wanted to ask whether there is a possibility for getting a test that checks the new code?

kronbichler commented:
I checked the PR and it looks good to me. Besides a test that would be highly desirable - simply apply the method to a (small) vector and output the result - may I also ask you to put an entry in `doc/news/changes.h` in the 'Specific Improvements' section? Thanks!

jgedicke commented:
As Guido pointed out the PreconditionBlock class does not support overlapping preconditioners as already mentioned in the description of that class, so I reverted those changes. Leaving the one line change in relaxation block preconditioners.
I added a test, a modification of solver_relaxation_03 to test for a nonsymmetric matrix and with small overlapping if the result changes when we reorder the blocks. For Jacobi the resulting vectors should of cause be equal.
But without this fix they are not (this is how I detected the bug).
I added a entry in `doc/news/change.h` in the specific improvements section.
Thanks!

kronbichler commented on doc/news/changes.h:
```diff
@@ -233,6 +233,14 @@ inconvenience this causes.
 <h3>Specific improvements</h3>
 
 <ol>
+ <li> Fixed: Overlapping block Jacobi. This fixes a bug in the RelaxationBlock
```
Could you please remove the "Overlapping block Jacobi" part? We should have the news message as short as possible and I think that the remainder explains it perfectly (it is also not the usual style in other messages).

jgedicke replied:
done!
Thx

kronbichler commented:
Well done! Thanks for the fix and the test case.
